File: go/registry.go
// registry.go
//
// ModelEntry + ModelRegistry + ModelRegistryService.
// CheckForUpgrades walks installed.json files in a storage directory
// and compares against the active registry, returning UpgradeInfo per drift.
//
// WriteInstalledManifest stamps installed.json after a successful install.

package circleai

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// ModelEntry is one model in the catalog.
type ModelEntry struct {
	Name         string       `json:"name"`
	Version      string       `json:"version"`
	Quantization string       `json:"quantization,omitempty"`
	URL          string       `json:"url,omitempty"`
	Checksum     string       `json:"checksum,omitempty"`
	Repo         string       `json:"repo,omitempty"`
	TotalBytes   int64        `json:"total_bytes,omitempty"`
	BundleFiles  []BundleFile `json:"bundle_files,omitempty"`
	MinRAMGB     float64      `json:"min_ram_gb,omitempty"`
	MinStorageGB float64      `json:"min_storage_gb,omitempty"`
	Capabilities []string     `json:"capabilities,omitempty"`
	QualityRank  int          `json:"quality_rank,omitempty"`
}

// IsBundle reports whether this entry is a multi-file bundle.
func (m ModelEntry) IsBundle() bool {
	return len(m.BundleFiles) > 0
}

// ModelRegistry is the active catalog snapshot.
type ModelRegistry struct {
	RegistryURL string       `json:"registry_url"`
	LastUpdated time.Time    `json:"last_updated"`
	Models      []ModelEntry `json:"models"`
}

// ModelRegistryService holds the active registry and exposes the upgrade detector.
type ModelRegistryService struct {
	registry *ModelRegistry
}

// NewModelRegistryService builds a service. Use SetRegistry to inject a registry.
func NewModelRegistryService() *ModelRegistryService {
	return &ModelRegistryService{}
}

// SetRegistry injects a registry (mainly for tests / synchronous workflows).
func (s *ModelRegistryService) SetRegistry(reg ModelRegistry) {
	s.registry = &reg
}

// AllModels returns the current registry contents.
func (s *ModelRegistryService) AllModels() []ModelEntry {
	if s.registry == nil {
		return nil
	}
	return s.registry.Models
}

// GetLatestModel returns the entry whose Name matches `modelName`, case-insensitive.
func (s *ModelRegistryService) GetLatestModel(modelName string) (ModelEntry, bool) {
	if s.registry == nil || modelName == "" {
		return ModelEntry{}, false
	}
	low := strings.ToLower(modelName)
	for _, m := range s.registry.Models {
		if strings.ToLower(m.Name) == low {
			return m, true
		}
	}
	return ModelEntry{}, false
}

// CheckForUpgrades walks every installed model under storageDirectory and
// returns one UpgradeInfo per detected drift (Version, file SHA, or both).
func (s *ModelRegistryService) CheckForUpgrades(storageDirectory string) ([]UpgradeInfo, error) {
	if storageDirectory == "" {
		return nil, errors.New("storageDirectory is required")
	}
	now := time.Now().UTC()
	out := make([]UpgradeInfo, 0)

	for _, entry := range s.AllModels() {
		modelDir := filepath.Join(storageDirectory, entry.Name)
		if !isDir(modelDir) {
			continue
		}
		manifestPath := filepath.Join(modelDir, "installed.json")
		manifest, ok := readManifest(manifestPath)
		if !ok {
			out = append(out, UpgradeInfo{
				ModelID:                entry.Name,
				InstalledVersion:       "",
				AvailableVersion:       entry.Version,
				Reason:                 UpgradeReasonUnknown,
				EstimatedDownloadBytes: entry.TotalBytes,
				DetectedAt:             now,
			})
			continue
		}

		versionChanged := manifest.Version != entry.Version
		shaChanged, driftBytes := compareBundleSha(manifest.Files, entry.BundleFiles)
		if !versionChanged && !shaChanged {
			continue
		}

		var reason UpgradeReason
		switch {
		case versionChanged && shaChanged:
			reason = UpgradeReasonBoth
		case versionChanged:
			reason = UpgradeReasonVersionChanged
		default:
			reason = UpgradeReasonShaChanged
		}

		out = append(out, UpgradeInfo{
			ModelID:                entry.Name,
			InstalledVersion:       manifest.Version,
			AvailableVersion:       entry.Version,
			Reason:                 reason,
			EstimatedDownloadBytes: driftBytes,
			DetectedAt:             now,
		})
	}

	return out, nil
}

// WriteInstalledManifest stamps installed.json into `modelDir`.
// Best-effort — silent failures so a manifest hiccup never breaks an install.
func WriteInstalledManifest(modelDir, modelID, version, repo string, bundleFiles []BundleFile) error {
	if modelDir == "" || modelID == "" {
		return errors.New("modelDir and modelID are required")
	}
	if err := os.MkdirAll(modelDir, 0o755); err != nil {
		return err
	}
	var total int64
	for _, f := range bundleFiles {
		if f.SizeBytes > 0 {
			total += f.SizeBytes
		}
	}
	m := InstalledManifest{
		ModelID:        modelID,
		Version:        version,
		Repo:           repo,
		TotalBytes:     total,
		Files:          bundleFiles,
		InstalledAtUTC: time.Now().UTC(),
	}
	b, err := json.MarshalIndent(m, "", "  ")
	if err != nil {
		return fmt.Errorf("marshal manifest: %w", err)
	}
	return os.WriteFile(filepath.Join(modelDir, "installed.json"), b, 0o644)
}

// ── helpers ─────────────────────────────────────────────────────────────

func isDir(p string) bool {
	info, err := os.Stat(p)
	return err == nil && info.IsDir()
}

func readManifest(p string) (InstalledManifest, bool) {
	b, err := os.ReadFile(p)
	if err != nil {
		return InstalledManifest{}, false
	}
	var m InstalledManifest
	if err := json.Unmarshal(b, &m); err != nil {
		return InstalledManifest{}, false
	}
	return m, true
}

func compareBundleSha(installed, available []BundleFile) (drift bool, bytes int64) {
	if len(available) == 0 {
		return false, 0
	}
	byName := make(map[string]BundleFile, len(installed))
	for _, f := range installed {
		byName[f.Name] = f
	}
	for _, av := range available {
		inst, found := byName[av.Name]
		if !found || !strings.EqualFold(inst.Sha256, av.Sha256) {
			drift = true
			bytes += av.SizeBytes
		}
	}
	return drift, bytes
}

File: rust/src/registry.rs
//! registry.rs
//!
//! ModelRegistryService + check_for_upgrades + write_installed_manifest.
//!
//! This is the byte-for-byte parity port of C# ModelRegistryService.CheckForUpgradesAsync.

use chrono::Utc;
use std::collections::HashMap;
use std::path::{Path, PathBuf};

use crate::catalog::{ModelEntry, ModelRegistry, ModelScopeCatalogClient};
use crate::models_v15::{BundleFile, InstalledManifest, UpgradeInfo, UpgradeReason};

pub struct ModelRegistryService {
    registry: Option<ModelRegistry>,
}

impl ModelRegistryService {
    pub fn new() -> Self {
        Self { registry: None }
    }

    pub fn from_catalog(client: &ModelScopeCatalogClient) -> Self {
        Self { registry: client.load_from_disk() }
    }

    pub fn set_registry(&mut self, reg: ModelRegistry) {
        self.registry = Some(reg);
    }

    pub fn all_models(&self) -> Vec<ModelEntry> {
        self.registry
            .as_ref()
            .map(|r| r.models.clone())
            .unwrap_or_default()
    }

    pub fn get_latest_model(&self, name: &str) -> Option<ModelEntry> {
        let lname = name.to_ascii_lowercase();
        self.registry
            .as_ref()?
            .models
            .iter()
            .find(|m| m.name.to_ascii_lowercase() == lname)
            .cloned()
    }

    /// Walks the storage dir and emits an `UpgradeInfo` for every installed
    /// model whose manifest is missing or drifts from the catalog.
    pub fn check_for_upgrades(&self, storage_directory: &Path) -> Vec<UpgradeInfo> {
        assert!(
            !storage_directory.as_os_str().is_empty(),
            "storage_directory must not be empty"
        );
        let now = Utc::now();
        let mut out = Vec::new();

        for entry in self.all_models() {
            let model_dir = storage_directory.join(&entry.name);
            if !model_dir.is_dir() {
                continue;
            }
            let manifest_path = model_dir.join("installed.json");
            let manifest = read_manifest(&manifest_path);
            let Some(m) = manifest else {
                out.push(UpgradeInfo {
                    model_id: entry.name.clone(),
                    installed_version: None,
                    available_version: entry.version.clone(),
                    reason: UpgradeReason::Unknown,
                    estimated_download_bytes: entry.total_bytes as i64,
                    detected_at: now,
                });
                continue;
            };

            let version_changed = m.version != entry.version;
            let (sha_changed, drift_bytes) = compare_bundle_sha(&m.files, &entry.bundle_files);
            if !version_changed && !sha_changed {
                continue;
            }
            let reason = match (version_changed, sha_changed) {
                (true, true) => UpgradeReason::Both,
                (true, false) => UpgradeReason::VersionChanged,
                (false, true) => UpgradeReason::ShaChanged,
                (false, false) => unreachable!(),
            };
            out.push(UpgradeInfo {
                model_id: entry.name.clone(),
                installed_version: Some(m.version.clone()),
                available_version: entry.version.clone(),
                reason,
                estimated_download_bytes: drift_bytes,
                detected_at: now,
            });
        }
        out
    }
}

impl Default for ModelRegistryService {
    fn default() -> Self {
        Self::new()
    }
}

fn read_manifest(path: &Path) -> Option<InstalledManifest> {
    let bytes = std::fs::read(path).ok()?;
    serde_json::from_slice(&bytes).ok()
}

fn compare_bundle_sha(installed: &[BundleFile], available: &[BundleFile]) -> (bool, i64) {
    if available.is_empty() {
        return (false, 0);
    }
    let by_name: HashMap<&str, &BundleFile> =
        installed.iter().map(|f| (f.name.as_str(), f)).collect();
    let mut drift = false;
    let mut bytes: i64 = 0;
    for av in available {
        match by_name.get(av.name.as_str()) {
            None => {
                drift = true;
                bytes += av.size_bytes;
            }
            Some(inst) if !inst.sha256.eq_ignore_ascii_case(&av.sha256) => {
                drift = true;
                bytes += av.size_bytes;
            }
            _ => {}
        }
    }
    (drift, bytes)
}

/// Writes installed.json after a successful model install. Best-effort —
/// silently no-ops on IO error so a failed write never breaks an otherwise
/// successful install.
pub fn write_installed_manifest(
    model_dir: &Path,
    model_id: &str,
    version: &str,
    repo: Option<&str>,
    bundle_files: Vec<BundleFile>,
) {
    let _ = (|| -> std::io::Result<()> {
        std::fs::create_dir_all(model_dir)?;
        let total: i64 = bundle_files.iter().map(|f| f.size_bytes.max(0)).sum();
        let manifest = InstalledManifest {
            model_id: model_id.to_string(),
            version: version.to_string(),
            repo: repo.map(|s| s.to_string()),
            total_bytes: total,
            files: bundle_files,
            installed_at_utc: Utc::now(),
        };
        let bytes = serde_json::to_vec_pretty(&manifest)
            .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))?;
        let path = model_dir.join("installed.json");
        std::fs::write(path, bytes)
    })();
}

pub fn _path_join(a: impl AsRef<Path>, b: impl AsRef<Path>) -> PathBuf {
    a.as_ref().join(b)
}
